### DISCUSSION TITLE
Parse Error when trying to run HelperConfig test

### DISCUSSION THREAD

**Discussion by neo-t-maredi** [Cyfrin/foundry-full-course-cu#1682] [Q&A] [answered]
Hey guys here's the error I'm getting 
![Screenshot 2024-05-04 161230](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/a5a386c1-a045-44ae-92bc-c5de1db6f87c)

This is the code I have and also there's a problem/Error with the helperConfig object itself. Help!! 
![Screenshot 2024-05-04 155632](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/e4f976ac-fdd2-495d-bb70-aa0bc4ec7b3a)
![Screenshot 2024-05-04 160118](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/dcb4cf06-b8bf-42c8-97ae-2092645e8461)
![Screenshot 2024-05-04 160131](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/602e8850-e640-441b-bd2d-b91e6b05c06a)

**Comment by EngrPips** [comment]
Hello @neo-t-maredi, on line 37 of your `HelperConfig.s.sol` file, your code should be written as I wrote below instead of as you wrote it.

```javascript
function getAnvilEthConfig() public pure returns(NetworkConfig memory) {
// You code logic here.
}
```

**Reply by neo-t-maredi** [reply]
Hey, thanks for the reply!! 
![Screenshot 2024-05-05 070129](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/e3a90590-d17d-492c-a108-c75231bc8c09)
![Screenshot 2024-05-05 070228](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/b5800b7a-f7b1-4cfa-b21b-390cf40c15ff)


Now I've corrected that and the error being thrown is: I feel like I'm losing my mind mat XD

**Reply by EngrPips** [reply]
Hello @neo-t-maredi, Please arrange your code in the order below.

```javascript
struct NetworkConfig {
address priceFeed;
}

NetworkConfig public activeNetworkConfig;
```

**Reply by neo-t-maredi** [reply]
Hey buddy, 

It's still just not working at all. I've tried forums and other avenues and it's been two days. This is beyond frustrating that it's just not working at all. 

I re-watched and tried coding it again line for line.
Maybe it's the extensions I have? What do you recommend? I should have been done with this a fortnight ago mate :/ 

![Screenshot 2024-05-07 132751](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/4ac0c56a-b338-43d1-a4b8-a6aa04e9bb9b)
![Screenshot 2024-05-07 132811](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/473cab67-e96f-46f8-aa6c-47058e9e8130)

**Reply by neo-t-maredi** [reply]
PS: Above is the change you recommended. Once implemented that's the error that was thrown subsequently

**Reply by neo-t-maredi** [reply]
DeployFundMe.s.sol

// SPDX-License-Identifier: MIT

pragma solidity ^0.8.19;

import {Script} from "lib/forge-std/src/Script.sol";
import {FundMe} from "../src/FundMe.sol";
import {HelperConfig} from "./HelperConfig.s.sol";

contract DeployFundMe is Script{
    function run() external returns (FundMe) {

        // Before startBroadcast -> Not a real transaction
        // Remove the declaration of the unused variable 'helperconfig'
        address ethUsdPriceFeed = helperConfig.activeNetworkConfig();
        vm.startBroadcast();
        //Mock contract to test veritability of test on different chains
        //FundMe fundMe = new FundMe(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        FundMe fundMe = new FundMe(ethUsdPriceFeed);
        vm.stopBroadcast();
        return fundMe;
    }
}

**Reply by neo-t-maredi** [reply]
HelperConfig.s.sol 

// SPDX-License-Identifier: MIT

// 1. Deploy mock contracts when we are on a local anvil chain 
// 2. Keep track of contract address across different chains 
// Sepolia ETH/USD 
// Mainnet ETH/USD 

pragma solidity ^0.8.19;

import {Script} from "lib/forge-std/src/Script.sol";

contract HelperConfig {
    // If we are on a local anvil, we deploy mocks; 
    // Otherwise, grab the existing address from the live network 
    struct NetWorkConfig {
        address priceFeed; //ETH/USD price feed address
        
    }
    
    NetworkConfig public activeNetworkConfig;

    constructor() {
        if (block.chainid == 11155111){
            activeNetworkConfig = getSepoliaEthConfig();
        }else {
            activeNetworkConfig = getAnvilConfig();
        }
    }
    function getSepoliaEthConfig() public pure returns (NetworkConfig memory){
        // price feed address 
        NetworkConfig memory sepoliaConfig = NetworkConfig({
            priceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306
        });
        return sepoliaConfig;
    }

    function getAnvilEthConfig() public pure returns(NetworkConfig memory){
        //price feed address
    }
}

**Reply by neo-t-maredi** [reply]
FundMeTest.t.sol

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {Test, console} from "lib/forge-std/src/Test.sol";
import {FundMe} from "../src/FundMe.sol";
import {DeployFundMe} from "../script/DeployFundMe.s.sol";

contract FundMeTest is Test {
    FundMe fundMe;

    function setUp() external {
        //fundMe = new FundMe(0x694AA1769357215DE4FAC081bf1f309aDC325306);
        DeployFundMe deployFundMe = new DeployFundMe();
        fundMe = deployFundMe.run();
    }

    function testMinimumDollarIsFive() public {
        assertEq(fundMe.MINIMUM_USD(), 5e18);
    }

        function testOwnerIsMsgSender() public {
        assertEq(fundMe.i_owner(), msg.sender);
    }
        //Testing nets with addresses outside our system?
        // 1. Unit: Testing a single function
        // 2. Integration: Testing multiple functions
        // 3. Forked: Testing on a forked network
        // 4. Staging: Testing on a live network (testnet or mainnet)
        function testPriceFeedVersionIsAccurate() public {
            uint256 version = fundMe.getVersion();
            assertEq(version, 4);
        }
         //Modular deployments
        //Modular testing

}

**Reply by neo-t-maredi** [reply]
Is there anything that I can do to fix this, I see know one else has had this problem at this stage. Tips please, even GPT4 is flummoxed.

**Reply by EngrPips** [reply]
```javascript
DeployFundMe.s.sol

// SPDX-License-Identifier: MIT

pragma solidity ^0.8.19;

import {Script} from "lib/forge-std/src/Script.sol";
import {FundMe} from "../src/FundMe.sol";
import {HelperConfig} from "./HelperConfig.s.sol";

contract DeployFundMe is Script{
function run() external returns (FundMe) {

    // Before startBroadcast -> Not a real transaction
    // Remove the declaration of the unused variable 'helperconfig'
    address ethUsdPriceFeed = helperConfig.activeNetworkConfig();
    vm.startBroadcast();
    //Mock contract to test veritability of test on different chains
    //FundMe fundMe = new FundMe(0x694AA1769357215DE4FAC081bf1f309aDC325306);
    FundMe fundMe = new FundMe(ethUsdPriceFeed);
    vm.stopBroadcast();
    return fundMe;
}
}
```

**Reply by EngrPips** [reply]
```js
HelperConfig.s.sol

// SPDX-License-Identifier: MIT

// 1. Deploy mock contracts when we are on a local anvil chain
// 2. Keep track of contract address across different chains
// Sepolia ETH/USD
// Mainnet ETH/USD

pragma solidity ^0.8.19;

import {Script} from "lib/forge-std/src/Script.sol";

contract HelperConfig {
// If we are on a local anvil, we deploy mocks;
// Otherwise, grab the existing address from the live network
struct NetWorkConfig {
address priceFeed; //ETH/USD price feed address

}

NetworkConfig public activeNetworkConfig;

constructor() {
    if (block.chainid == 11155111){
        activeNetworkConfig = getSepoliaEthConfig();
    }else {
        activeNetworkConfig = getAnvilConfig();
    }
}
function getSepoliaEthConfig() public pure returns (NetworkConfig memory){
    // price feed address 
    NetworkConfig memory sepoliaConfig = NetworkConfig({
        priceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306
    });
    return sepoliaConfig;
}

function getAnvilEthConfig() public pure returns(NetworkConfig memory){
    //price feed address
}
}
```

**Reply by EngrPips** [reply]
```js
FundMeTest.t.sol

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {Test, console} from "lib/forge-std/src/Test.sol";
import {FundMe} from "../src/FundMe.sol";
import {DeployFundMe} from "../script/DeployFundMe.s.sol";

contract FundMeTest is Test {
FundMe fundMe;

function setUp() external {
    //fundMe = new FundMe(0x694AA1769357215DE4FAC081bf1f309aDC325306);
    DeployFundMe deployFundMe = new DeployFundMe();
    fundMe = deployFundMe.run();
}

function testMinimumDollarIsFive() public {
    assertEq(fundMe.MINIMUM_USD(), 5e18);
}

    function testOwnerIsMsgSender() public {
    assertEq(fundMe.i_owner(), msg.sender);
}
    //Testing nets with addresses outside our system?
    // 1. Unit: Testing a single function
    // 2. Integration: Testing multiple functions
    // 3. Forked: Testing on a forked network
    // 4. Staging: Testing on a live network (testnet or mainnet)
    function testPriceFeedVersionIsAccurate() public {
        uint256 version = fundMe.getVersion();
        assertEq(version, 4);
    }
     //Modular deployments
    //Modular testing
}
```

**Reply by EngrPips** [reply, marked as answer]
Hello @neo-t-maredi, In your `HelperConfig.s.sol` why did you define `NetWorkConfig` with capital letter `W` in your spelling as shown in the below code.

```
struct NetWorkConfig {
address priceFeed; //ETH/USD price feed address

}
```
But when you wanted to declare the type you declared `NetworkConfig` with small letter `w` in your spelling as shown below.

```js
NetworkConfig public activeNetworkConfig;
```

**Reply by neo-t-maredi** [reply]
Ka-CHOW!!!! It worked :) Sorry for delayed response been super busy at calendar year end at work. Thanks so much mate!!

**Reply by EngrPips** [reply]
Awesome! Keep crushing it, friend.

**Reply by neo-t-maredi** [reply]
Hey Friend 

I'm currently on lesson/part 13 of the Foundry Fundamentals course. I'm currently having an error (7576) thrown as below:

![Screenshot 2024-05-13 092229](https://github.com/Cyfrin/foundry-full-course-f23/assets/146746647/d2b52eb4-e077-425b-bb4e-77af0981a69a)

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
    using PriceConverter for uint256;

    mapping(address => uint256) public s_addressToAmountFunded;
    address[] public s_funders;

    // Could we make this constant?  /* hint: no! We should make it immutable! */
    address public /* immutable */ i_owner;
    uint256 public constant MINIMUM_USD = 5e18;
    AggregatorV3Interface private s_priceFeed;
    
    constructor(address priceFeed) {
        i_owner = msg.sender;
        s_priceFeed = AggregatorV3Interface(priceFeed);

    }

    function fund() public payable {
        require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
        // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
        s_addressToAmountFunded[msg.sender] += msg.value;
        s_funders.push(msg.sender);
    }
    
    function getVersion() public view returns (uint256){
       return s_priceFeed.version();
    }
    
    modifier onlyOwner {
        // require(msg.sender == owner);
        if (msg.sender != i_owner) revert FundMe__NotOwner();
        _;
    }
    
    function withdraw() public onlyOwner {
        for (
            uint256 funderIndex=0; 
            funderIndex < s_funders.length; 
            funderIndex++
        ) {
            address funder = s_funders[funderIndex];
            s_addressToAmountFunded[funder] = 0;
        }
        s_funders = new address[](0);
        // // transfer
        // payable(msg.sender).transfer(address(this).balance);
        
        // // send
        // bool sendSuccess = payable(msg.sender).send(address(this).balance);
        // require(sendSuccess, "Send failed");

        // call
        (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
        require(callSuccess, "Call failed");
    }
    // Explainer from: https://solidity-by-example.org/fallback/
    // Ether is sent to contract
    //      is msg.data empty?
    //          /   \ 
    //         yes  no
    //         /     \
    //    receive()?  fallback() 
    //     /   \ 
    //   yes   no
    //  /        \
    //receive()  fallback()

    fallback() external payable {
        fund();
    }

    receive() external payable {
        fund();
    }

    /**
     * View / Pure functions (Getters)
     */

     function getAddressToAmountFunded(
        address fundingAddress
    ) external view returns (uint256) {
         return s_addressToAmountFunded[fundingAddress];
     }

     function getFunder() external view returns (address) {
         return s_funders[index];
     }
}


I've enabled co-pilot and haven't seen any errors when writing the initial code but running into this afterwards.

**Reply by EngrPips** [reply]
```js
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.19;

import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";
import {PriceConverter} from "./PriceConverter.sol";

error FundMe__NotOwner();

contract FundMe {
using PriceConverter for uint256;

mapping(address => uint256) public s_addressToAmountFunded;
address[] public s_funders;

// Could we make this constant?  /* hint: no! We should make it immutable! */
address public /* immutable */ i_owner;
uint256 public constant MINIMUM_USD = 5e18;
AggregatorV3Interface private s_priceFeed;

constructor(address priceFeed) {
    i_owner = msg.sender;
    s_priceFeed = AggregatorV3Interface(priceFeed);

}

function fund() public payable {
    require(msg.value.getConversionRate(s_priceFeed) >= MINIMUM_USD, "You need to spend more ETH!");
    // require(PriceConverter.getConversionRate(msg.value) >= MINIMUM_USD, "You need to spend more ETH!");
    s_addressToAmountFunded[msg.sender] += msg.value;
    s_funders.push(msg.sender);
}

function getVersion() public view returns (uint256){
   return s_priceFeed.version();
}

modifier onlyOwner {
    // require(msg.sender == owner);
    if (msg.sender != i_owner) revert FundMe__NotOwner();
    _;
}

function withdraw() public onlyOwner {
    for (
        uint256 funderIndex=0; 
        funderIndex < s_funders.length; 
        funderIndex++
    ) {
        address funder = s_funders[funderIndex];
        s_addressToAmountFunded[funder] = 0;
    }
    s_funders = new address[](0);
    // // transfer
    // payable(msg.sender).transfer(address(this).balance);
    
    // // send
    // bool sendSuccess = payable(msg.sender).send(address(this).balance);
    // require(sendSuccess, "Send failed");

    // call
    (bool callSuccess, ) = payable(msg.sender).call{value: address(this).balance}("");
    require(callSuccess, "Call failed");
}
// Explainer from: https://solidity-by-example.org/fallback/
// Ether is sent to contract
//      is msg.data empty?
//          /   \ 
//         yes  no
//         /     \
//    receive()?  fallback() 
//     /   \ 
//   yes   no
//  /        \
//receive()  fallback()

fallback() external payable {
    fund();
}

receive() external payable {
    fund();
}

/**
 * View / Pure functions (Getters)
 */

 function getAddressToAmountFunded(
    address fundingAddress
) external view returns (uint256) {
     return s_addressToAmountFunded[fundingAddress];
 }

 function getFunder() external view returns (address) {
     return s_funders[index];
 }
}
```

**Reply by EngrPips** [reply]
Hello @neo-t-maredi, where did you get this `index` variable you used in the code below?

```js
function getFunder() external view returns (address) {
     return s_funders[index];
 }
```
I think the correct code should be the one below 

```js
function getFunder(uint256 index) external view returns (address) {
     return s_funders[index];
 }
```

**Reply by neo-t-maredi** [reply]
Ah when I first put on the parameters I put a commas and then removed the variables in the parameter. My bad. Thanks so much buddy, rookie error. Thanks for your patience

**Reply by EngrPips** [reply]
Anytime, Friend. I am glad I could help.
